relay/network_graph/dijkstra_weighted.py

```diff
@@ -111,7 +46,68 @@ def find_path(G, source, target, get_fee, value, max_hops=None, max_fees=None):
     except KeyError:
         raise nx.NetworkXNoPath(
             "node %s not reachable from %s" % (source, target))
+
+
+def find_path_triangulation(G, source, target_reduce, target_increase, get_fee, value, max_hops=None, max_fees=None):
+    """
+    target_reduce is the node we want to reduce our credit with (for value > 0)
+    target_increase is the node which will result with an increased debt (for value > 0)
+    source will owe less to target_reduce and target_increase will owe more to source (for value > 0)
+    must return a path and the total fee of that path
+    to be called in the right order with source as source and target as target, contrary to find_path
+
+    value may be negative, in which case
+    source will owe more to target_reduce and target_increase will owe less to source
+    """
+    def get_fee_wrapper(b, a, value):
+        # used to get the data from the graph in the right order and query the fees
+        if b < a:
+            return get_fee(b, a, G[b][a], value)
+        else:
+            return get_fee(b, a, G[a][b], value)
+
+    def verify_balance_greater_than_value(a, b, value):
+        # used to verify that we reduce the amount source owes to target_reduce and do not misuse the function
+        if a < b:
+            return -G[a][b]['balance_ab'] >= value
+        else:
+            return G[b][a]['balance_ab'] >= value
+
+    def output(final_fee, intermediary_path):
+        if max_fees is not None and final_fee > max_fees:
+            raise nx.NetworkXNoPath(
+                "operation impossible due to fees: %d with max_fees: %d" % (final_fee, max_fees))
+        return (final_fee, list(reversed([source] + intermediary_path + [source])))
+
+    # verification that the funtion is used properly
+    if value <= 0:
+        raise ValueError('This value cannot be handled yet : %d' % value)
+    elif not verify_balance_greater_than_value(source, target_reduce, value):
+        raise ValueError('The value is greater than the amount source owes target_reduce : %d' % value)
+
+    G_succ = G.succ if G.is_directed() else G.adj
+    neighbors = [x[0] for x in G_succ[source].items()]
+
+    if target_reduce not in neighbors:
+        raise nx.NetworkXNoPath(
+            "node %s not a neighbor of %s" % (target_reduce, source))
+    if target_increase not in neighbors:
+        raise nx.NetworkXNoPath(
+            "node %s not a neighbor of %s" % (target_increase, source))
+
+    first_fee = get_fee_wrapper(source, target_reduce, value)
+
+    if max_hops is not None:
+        max_hops -= 2
+
+    intermediary_fee, intermediary_path = find_path(G, target_reduce, target_increase,
+                                                    get_fee, value+first_fee, max_hops, max_fees, source)
+
+    last_fee = get_fee_wrapper(target_increase, source, value+intermediary_fee+first_fee)
+
+    final_fee = last_fee + intermediary_fee + first_fee
+    return output(final_fee, intermediary_path)
```

Comment:
I'd say output now suffers a little bit from that problem that you had with functions, that it is unclear what output is doing from looking at its call. Maybe come up with a more descriptive name or dont use a function there.